apps/04TowersOfHanoi.js

```diff
@@ -10,74 +10,209 @@ var stacks = {
 }
 
-function movePiece() {
-    // Your code here
+function printNumberOfTurns(numberOfTurns) {
+    console.log("# of Turns: " + numberOfTurns);
+}
+
+function movePiece(startStack, endStack) {
+
+    var lengthA = stacks.a.length;
+    var lengthB = stacks.b.length;
+    var lengthC = stacks['c'].length;
+
+    //adjust the array to move the last number from the start stack and add to the front of the end stack
+    var lastValueA = stacks.a[lengthA - 1];
+    var lastValueB = stacks.b[lengthB - 1];
+    var lastValueC = stacks.c[lengthC - 1];
+
+    if (startStack === 'a') {
+        stacks.a.pop();
+
+        if (endStack === 'b') {
+            stacks.b.push(lastValueA);
+        } else {
+            stacks.c.push(lastValueA);
+        }
+
+    } else if (startStack === 'b') {
+        stacks.b.pop();
+
+        if (endStack === 'a') {
+            stacks.a.push(lastValueB);
+        } else {
+            stacks.c.push(lastValueB);
+        }
+
+    } else {
+        stacks.c.pop();
+
+        if (endStack === 'a') {
+            stacks.a.push(lastValueC);
+        } else {
+            stacks.b.push(lastValueC);
+        }
+
+    }
 
 }
 
-function isLegal() {
-    // Your code here
+function isLegal(startStack, endStack) {
+
+    var lengthStartStack = stacks[startStack].length;
+    var lengthEndStack = stacks[endStack].length;
+
+    var lastValueStartStack = stacks[startStack][lengthStartStack - 1];
+    var lastValueEndStack = stacks[endStack][lengthEndStack - 1];
+
+    if (lengthStartStack === 0) {
+        lastValueStartStack = 0;
+    }
+    if (lengthEndStack === 0) {
+        lastValueEndStack = 0;
+    }
+
+    // move is legal if the value you are moving to the endStack is LESS than the last value of the startStack
+    console.log('lastValueStartStack = ' + lastValueStartStack + ' and lastValueEndStack = ' + lastValueEndStack)
+
+    if (lastValueEndStack === 0) {
+        return true;
+    } else if (lastValueStartStack > lastValueEndStack) {
+        return false;
+    } else if (lastValueStartStack < lastValueEndStack) {
+        return true;
+    }
 
 }
 
 function checkForWin() {
-    // Your code here
+
+    //the game is won when either stack b or stack c has all 4 in the correct order
+
+    var stackAsStringB = stacks['b'].toString();
+    var stackAsStringC = stacks['c'].toString();
+
+    if (stackAsStringB === '4,3,2,1' || stackAsStringC === '4,3,2,1') {
+        return true;
+    } else {
+        return false;
+    }
 
 }
 
 function towersOfHanoi(startStack, endStack) {
-    // Your code here
+    // FIRST check isLegal
+    // If isLegal is true, then do movePiece.  If not true, ask for prompt again.
+    // THIRD checkForWin - if win, end game.  if no win, continue function.
+    // FOURTH if no winner, increase numberOfTurns & continue game
+
+    var isLegalFunction = isLegal(startStack, endStack);
+
+    console.log('Is the move legal? ' + isLegalFunction);
+
+    if (isLegalFunction === true) {
```

Comment:
Checking === true isn't usually necessary. What datatype is the variable `isLegalFunction` ?